index.go

```diff
@@ -1,47 +1,2085 @@
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
 	"github.com/google/go-containerregistry/pkg/v1/types"
+	"golang.org/x/sync/errgroup"
 )
 
 type ImageIndex interface {
 	// getters
 
-	Name() string
+	OS(digest name.Digest) (os string, err error)
+	Architecture(digest name.Digest) (arch string, err error)
+	Variant(digest name.Digest) (osVariant string, err error)
+	OSVersion(digest name.Digest) (osVersion string, err error)
+	Features(digest name.Digest) (features []string, err error)
+	OSFeatures(digest name.Digest) (osFeatures []string, err error)
+	Annotations(digest name.Digest) (annotations map[string]string, err error)
+	URLs(digest name.Digest) (urls []string, err error)
 
-	// modifiers
-	Add(repoName string) error
-	Remove(repoName string) error
-	Save(additionalNames ...string) error
+	// setters
+
+	SetOS(digest name.Digest, os string) error
+	SetArchitecture(digest name.Digest, arch string) error
+	SetVariant(digest name.Digest, osVariant string) error
+	SetOSVersion(digest name.Digest, osVersion string) error
+	SetFeatures(digest name.Digest, features []string) error
+	SetOSFeatures(digest name.Digest, osFeatures []string) error
+	SetAnnotations(digest name.Digest, annotations map[string]string) error
+	SetURLs(digest name.Digest, urls []string) error
+
+	// misc
+
+	Add(ref name.Reference, ops ...IndexAddOption) error
+	Save() error
+	Push(ops ...IndexPushOption) error
+	Inspect() error
+	Remove(digest name.Digest) error
+	Delete() error
+}
+
+var (
+	ErrOSUndefined                        = errors.New("os is undefined")
+	ErrArchUndefined                      = errors.New("architecture is undefined")
+	ErrVariantUndefined                   = errors.New("variant is undefined")
+	ErrOSVersionUndefined                 = errors.New("osVersion is undefined")
+	ErrFeaturesUndefined                  = errors.New("features are undefined")
+	ErrOSFeaturesUndefined                = errors.New("os-features are undefined")
+	ErrURLsUndefined                      = errors.New("urls are undefined")
+	ErrAnnotationsUndefined               = errors.New("annotations are undefined")
+	ErrNoImageOrIndexFoundWithGivenDigest = errors.New("no image/index found with the given digest")
+	ErrConfigFilePlatformUndefined        = errors.New("platform is undefined in config file")
+	ErrManifestUndefined                  = errors.New("manifest is undefined")
+	ErrPlatformUndefined                  = errors.New("platform is undefined")
+	ErrInvalidPlatform                    = errors.New("invalid platform is provided")
+	ErrConfigFileUndefined                = errors.New("config file is undefined")
+	ErrIndexNeedToBeSaved                 = errors.New("image index should need to be saved to perform this operation")
+	ErrUnknownMediaType                   = errors.New("media type not supported")
+	ErrNoImageFoundWithGivenPlatform      = errors.New("no image found with the given platform")
+)
```

Comment:
@jjbustamante it would be great if you also suggest some changes to error messages